price-estimator/src/solver_rounding_buffer.rs

```diff
@@ -0,0 +1,162 @@
+        .external_price as f64;
+    assert!(fee_token_price > 0.0, "fee token price is 0");
+
+    // The maximum rounding buffer over all orders from this address selling this token.
+    let mut account_balance_buffers = HashMap::<(Address, TokenId), u128>::new();
+    // Apply rounding buffer to account balances and order sell amounts.
+    for o in orders.iter_mut() {
+        // Compute rounding buffer.
+        let (t_s, t_b) = (TokenId(o.sell_token), TokenId(o.buy_token));
+        let estimated_buy_token_price = match tokens.get(&t_b) {
+            Some(TokenInfo { external_price, .. }) if *external_price > 0 => *external_price as f64,
+            _ => continue,
+        };
+        let estimated_sell_token_price = match tokens.get(&t_s) {
+            Some(TokenInfo { external_price, .. }) if *external_price > 0 => *external_price as f64,
+            _ => continue,
+        };
+        let estimated_xrate = estimated_buy_token_price / estimated_sell_token_price;
+        let rounding_buffer = max_rounding_amount(estimated_buy_token_price, fee_token_price)
+            * estimated_xrate
+            * PRICE_ESTIMATION_ERROR.powi(2);
+        let rounding_buffer = rounding_buffer.ceil() as u128;
+
+        // Update rounding buffer for account balances;
+        let entry = account_balance_buffers
+            .entry((o.account_id, t_s))
+            .or_default();
+        *entry = (*entry).max(rounding_buffer);
+
+        // Reduce order sell amount.
+        o.denominator = o.denominator.saturating_sub(rounding_buffer);
+        o.remaining_sell_amount = o.remaining_sell_amount.saturating_sub(rounding_buffer);
+    }
+
+    // Reduce account balances.
+    for ((address, token_id), rounding_buffer) in account_balance_buffers {
+        if let Some(balance) = account_state.0.get_mut(&(address, token_id.0)) {
+            *balance = balance.saturating_sub(U256::from(rounding_buffer));
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ethcontract::{Address, U256};
+
+    fn address(n: u64) -> Address {
+        Address::from_low_u64_le(n)
+    }
+
+    fn account(address_: u64, token: u16, balance: u128) -> ((Address, u16), U256) {
+        ((address(address_), token), U256::from(balance))
+    }
+
+    fn token_info(external_price: u128) -> TokenInfo {
+        TokenInfo {
+            alias: None,
+            decimals: None,
+            external_price,
+        }
+    }
+
+    fn order(
+        id: u16,
+        address_: u64,
+        buy_token: u16,
+        sell_token: u16,
+        numerator: u128,
+        denominator: u128,
+    ) -> Order {
+        Order {
+            id,
+            account_id: address(address_),
+            buy_token,
+            sell_token,
+            numerator,
+            denominator,
+            remaining_sell_amount: denominator,
+            valid_from: 0,
+            valid_until: 0,
+        }
+    }
+
+    #[test]
+    fn apply_rounding_buffer_ok() {
```

Comment:
Is this test taken from the python codebase?

While it is nice, and reassures that the logic is correct, it doesn't really document the expected behaviour (e.g. that we take the max rounding per token to compute the rounding on the balance, that rounding depends on prices but not on amounts, etc.). Maybe we could add a few more documentary tests given that this is a quite custom and 1-off algortithm.

Reply:
I don't think it makes sense to test details of the algorithm. After all it has simply been copied from python and we will always keep it up to date with the python version. It's not the job of this code to verify that the python algorithm is sensible. Unless you think it would help catch bugs introduced by converting to rust.

I wrote this test by running the python function manually with the same inputs to get the expected result.